primer/brand
# Convert Primer Brand to a monorepo

## Summary

Conversion of Primer Brand to a monorepo [per the ADR](https://github.com/primer/brand/blob/main/contributor-docs/adrs/adr-002-conversion-to-monorepo.md). 

Part of https://github.com/github/primer/issues/1366

## List of notable changes:



- **added** the necessary configurations for an npm workspace
- **moved** `react`, `e2e`, `design-tokens` and `storybook` into their own packages 
- **published** new package for `@primer/brand-primitives`
- **added** custom workflow for publishing multiple packages
- **fixed** previously broken start commands, now opens both docs and storybook, while running css module type generator in the background

## Benchmarks

Before and after benchmarks for CI runs.

|                                          | Single repo | Monorepo |
|------------------------------------------|-------------|----------|
| CI / Install, lint, format, check and tests               | 2m          | 3m       |
| CodeQL / Analyze (javascript)            | 2m          | 1m       |
| Integration test / Astro - React starter | 4m          | 6m       |
| Integration test / Create React App      | 4m          | 8m       |
| Integration test / Next.js               | 6m          | 5m       |
| Integration test / Remix.run             | 3m          | 3m       |
| UI test / Accessibility                  | 3m          | 5m       |
| UI test / Primitives diff                | 4m          | 2m       |
| Visual regression tests                  | 4m          | 6m       |
| Preview / Build & Deploy                 | 5m 7s       | 9m 7s    |
| Release canary                           | 1m          | 1m       |



## What should reviewers focus on?

- Reviewing ergonomic differences in moving to a monorepo
- Checking that nothing critical has been missed out

## Steps to test:



1. Checkout branch `git checkout -b rezrah/npm-workspace origin/rezrah/npm-workspace`
2. `npm install && cd apps/docs && npm install --legacy-peer-deps` (because docs is outside workspace for now)
3. `npm run build`
4. `npm run lint`
4. `npm run format`
4. `npm run check`
4. `npm run test`
4. `npm run start` (should open Storybook, Docs in the browser)

## Supporting resources (related issues, external links, etc):

- NPM Workspaces https://docs.npmjs.com/cli/v8/using-npm/workspaces

## Contributor checklist:


## Reviewer checklist:


## Screenshots:
![brand-primitives and react-brand both published](https://user-images.githubusercontent.com/13340707/195093880-38ff7457-a2b4-4b4f-8e9b-35d15729fda4.png)
